cli/src/commands/test/gas_report.rs

```diff
@@ -0,0 +1,158 @@
+// Copyright 2023 Polyphene.
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use std::collections::BTreeMap;
+
+use comfy_table::{
+    modifiers::UTF8_ROUND_CORNERS, presets::UTF8_FULL, Attribute, Cell, Color, Table,
+};
+use kythera_lib::{ExecutionEvent, Method, TestResult, TestResultType, WasmActor};
+
+/// Gas report for the tested contracts.
+#[derive(Default, Debug)]
+pub struct GasReport<'a> {
+    reports: BTreeMap<&'a WasmActor, ActorInfo<'a>>,
+}
+
+#[derive(Debug, Default)]
+/// Actor method calls information
+/// TODO: calculate actor deployment gas consumption.
+pub struct ActorInfo<'a> {
+    methods: BTreeMap<&'a Method, Vec<u64>>,
+}
+
+impl<'a> GasReport<'a> {
+    /// Analyze a set of [`TestResult`]s for a target Actor.
+    pub fn analyze(&mut self, actor: &'a WasmActor, test_results: &[&TestResult]) {
+        let mut info = match self.reports.remove(actor) {
+            Some(info) => info,
+            None => ActorInfo::default(),
+        };
+
+        for result in test_results {
+            let apply_ret = match result.ret() {
+                TestResultType::Passed(apply_ret) | TestResultType::Failed(apply_ret) => apply_ret,
+                TestResultType::Erred(_) => {
+                    continue;
+                }
+            };
+
+            // Get the Gas consumption by each Call of the Target actor.
+            let mut stack = vec![];
+            for trace in &apply_ret.exec_trace {
+                match trace {
+                    ExecutionEvent::GasCharge(gas_charge) => {
+                        // Add this gas charge to the total gas charge by the main method.
+                        // There is a `GasCharge` before any `Call` so we skip it.
+                        let (_method, total_gas_charge) = match stack.get_mut(0) {
+                            Some(e) => e,
+                            None => continue,
+                        };
+                        *total_gas_charge =
+                            *total_gas_charge + gas_charge.compute_gas.as_milligas();
```

Comment:
I think the gas charge should be evolving for all calls currently being processed, as everything left in the stack are subsequent calls of each other.

Reply:
A chain of call that could happen and we still want to cover: 

`TestActor::TestMethod()` -> `Proxy::Relay()` -> `TargetActor::MethodA()` -> `OtherActor::OtherMethodA()` -> `TargetActor::MethodB()`

This means that the traces would be organized as such:

```
<Call `TestActor::TestMethod()` >
  <Call `Proxy::Relay()` >
    <Call `TargetActor::MethodA()` >
      <Call `OtherActor::OtherMethodA()` >
        <Call `TargetActor::MethodB()` >
        <CallReturn `TargetActor::MethodB()` >
      <CallReturn `OtherActor::OtherMethodA()` >
    <CallReturn `TargetActor::MethodA()` >
  <CallReturn `Proxy::Relay()` >
<CallReturn `TestActor::TestMethod()` >
```

Reply:
I guess that instead of writing on `stack.get_mut(0)` you should write on `stack.get_mut(stack.len - 1)`. 

But you would also need to pass a subsequent call gas value to the next call in the stack in the `ExecutionEvent::CallReturn(_, _) | ExecutionEvent::CallError(_)` branch.

Reply:
the `Vec` is being used as stack, therefore when we check that method call is from the target actor we push it and the total gas consumed which is 0 at that time.  
This means `stack[0] ` contains the target actor method call and total gas consumed. Every time there is a Call we push to the stack again, which means that the next (1,2,3) indices of the `Vec` will be filled.  
On each `ExecutionEvent::GasCharge` we check if we are inside the scope of a target actor method by calling `stack.get_mut(0)` and if it exists we add to the `total_gas_charge`. 
Then on each `ExecutionEvent::CallReturn` we pop from the stack and check if it's empty, means that we popped the target actor method and it's total gas consumption, and add those to the gas report. 
I believe this reflects the traces diagram you described above right?